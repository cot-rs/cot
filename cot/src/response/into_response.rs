use std::borrow::Cow;
use std::convert::Infallible;

use bytes::{Bytes, BytesMut};
use cot::response::Response;
use cot::{Body, StatusCode};
use http;

/// Trait for generating responses.
/// Types that implement `IntoResponse` can be returned from handlers.
///
/// # Implementing `IntoResponse`
///
/// You generally shouldn't have to implement `IntoResponse` manually, as cot
/// provides implementations for many common types.
///
/// However, it might be necessary if you have a custom error type that you want
/// to return from handlers.

pub trait IntoResponse {
    /// Create a response.
    #[must_use]
    fn into_response(self) -> Response;

    fn with_header<K, V>(self, key: K, value: V) -> Response
    where
        K: Into<http::HeaderName>,
        V: Into<http::HeaderValue>,
        Self: Sized,
    {
        let key = key.into();
        let value = value.into();

        let mut response = self.into_response();
        response.headers_mut().append(key, value);
        response
    }

    fn with_content_type<V>(self, content_type: V) -> Response
    where
        V: Into<http::HeaderValue>,
        Self: Sized,
    {
        let mut response = self.into_response();
        response
            .headers_mut()
            .insert(http::header::CONTENT_TYPE, content_type.into());
        response
    }

    fn with_status(self, status: StatusCode) -> Response
    where
        Self: Sized,
    {
        let mut response = self.into_response();
        *response.status_mut() = status;
        response
    }

    fn with_body(self, body: impl Into<Body>) -> Response
    where
        Self: Sized,
    {
        let mut response = self.into_response();
        *response.body_mut() = body.into();
        response
    }
}
macro_rules! impl_into_response_for_type_and_mime {
    ($ty:ty, $mime:expr) => {
        impl IntoResponse for $ty {
            fn into_response(self) -> Response {
                let mut res = Body::from(self).into_response();
                res.headers_mut().insert(
                    http::header::CONTENT_TYPE,
                    http::HeaderValue::from_static($mime.as_ref()),
                );
                res
            }
        }
    };
}

// General implementations

impl IntoResponse for () {
    fn into_response(self) -> Response {
        Body::empty().into_response()
    }
}

impl IntoResponse for Infallible {
    fn into_response(self) -> Response {
        match self {}
    }
}

impl<R, E> IntoResponse for Result<R, E>
where
    R: IntoResponse,
    E: IntoResponse,
{
    fn into_response(self) -> Response {
        match self {
            Ok(value) => value.into_response(),
            Err(err) => err.into_response(),
        }
    }
}

impl IntoResponse for Response {
    fn into_response(self) -> Response {
        self
    }
}
// Text implementations
impl_into_response_for_type_and_mime!(&'static str, mime::TEXT_PLAIN_UTF_8);
impl_into_response_for_type_and_mime!(String, mime::TEXT_PLAIN_UTF_8);

impl IntoResponse for Box<str> {
    fn into_response(self) -> Response {
        String::from(self).into_response()
    }
}

// Bytes implementations
impl_into_response_for_type_and_mime!(&'static [u8], mime::APPLICATION_OCTET_STREAM);
impl_into_response_for_type_and_mime!(Vec<u8>, mime::APPLICATION_OCTET_STREAM);
impl_into_response_for_type_and_mime!(Bytes, mime::APPLICATION_OCTET_STREAM);

impl<const N: usize> IntoResponse for &'static [u8; N] {
    fn into_response(self) -> Response {
        self.as_slice().into_response()
    }
}

impl<const N: usize> IntoResponse for [u8; N] {
    fn into_response(self) -> Response {
        self.to_vec().into_response()
    }
}

impl IntoResponse for Box<[u8]> {
    fn into_response(self) -> Response {
        Vec::from(self).into_response()
    }
}

impl IntoResponse for BytesMut {
    fn into_response(self) -> Response {
        self.freeze().into_response()
    }
}

// HTTP structures for common uses

impl IntoResponse for StatusCode {
    fn into_response(self) -> Response {
        let mut res = ().into_response();
        *res.status_mut() = self;
        res
    }
}

impl IntoResponse for http::HeaderMap {
    fn into_response(self) -> Response {
        let mut res = ().into_response();
        *res.headers_mut() = self;
        res
    }
}

impl IntoResponse for http::Extensions {
    fn into_response(self) -> Response {
        let mut res = ().into_response();
        *res.extensions_mut() = self;
        res
    }
}

impl IntoResponse for http::response::Parts {
    fn into_response(self) -> Response {
        Response::from_parts(self, Body::empty())
    }
}

// Shortcuts for common uses

impl IntoResponse for Body {
    fn into_response(self) -> Response {
        Response::new(self)
    }
}

impl<R> IntoResponse for (StatusCode, R)
where
    R: IntoResponse,
{
    fn into_response(self) -> Response {
        let mut res = self.1.into_response();
        *res.status_mut() = self.0;
        res
    }
}

impl<R> IntoResponse for (StatusCode, http::HeaderMap, R)
where
    R: IntoResponse,
{
    fn into_response(self) -> Response {
        let mut res = self.2.into_response();
        *res.status_mut() = self.0;
        *res.headers_mut() = self.1;
        res
    }
}

impl<R> IntoResponse for (http::HeaderMap, R)
where
    R: IntoResponse,
{
    fn into_response(self) -> Response {
        let mut res = self.1.into_response();
        *res.headers_mut() = self.0;
        res
    }
}
