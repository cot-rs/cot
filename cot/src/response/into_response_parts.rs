use crate::response::IntoResponse;

/// Trait for adding headers and extensions to a response.
pub trait IntoResponseParts {
    /// The type returned in the event of an error.
    ///
    /// This can be used to fallibly convert types into headers or extensions.
    type Error: IntoResponse;

    /// Set parts of the response
    fn into_response_parts(
        self,
        res: http::response::Parts,
    ) -> Result<http::response::Parts, Self::Error>;
}
