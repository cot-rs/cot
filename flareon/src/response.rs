use crate::headers::HTML_CONTENT_TYPE;
use crate::{Body, StatusCode};

const RESPONSE_BUILD_FAILURE: &str = "Failed to build response";

pub type Response = http::Response<Body>;

pub trait ResponseExt {
    #[must_use]
    fn new_html(status: StatusCode, body: Body) -> Self;

    #[must_use]
    fn new_redirect<T: Into<String>>(location: T) -> Self;
}

impl ResponseExt for Response {
    fn new_html(status: StatusCode, body: Body) -> Self {
        http::Response::builder()
            .status(status)
            .header(http::header::CONTENT_TYPE, HTML_CONTENT_TYPE)
            .body(body)
            .expect(RESPONSE_BUILD_FAILURE)
    }

    fn new_redirect<T: Into<String>>(location: T) -> Self {
        http::Response::builder()
            .status(StatusCode::SEE_OTHER)
            .header(http::header::LOCATION, location.into())
            .body(Body::empty())
            .expect(RESPONSE_BUILD_FAILURE)
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::headers::HTML_CONTENT_TYPE;
    use crate::response::{Response, ResponseExt};

    #[test]
    fn test_response_new_html() {
        let body = Body::fixed("<html></html>");
        let response = Response::new_html(StatusCode::OK, body);
        assert_eq!(response.status(), StatusCode::OK);
        assert_eq!(
            response.headers().get(http::header::CONTENT_TYPE).unwrap(),
            HTML_CONTENT_TYPE
        );
    }

    #[test]
    fn test_response_new_redirect() {
        let location = "http://example.com";
        let response = Response::new_redirect(location);
        assert_eq!(response.status(), StatusCode::SEE_OTHER);
        assert_eq!(
            response.headers().get(http::header::LOCATION).unwrap(),
            location
        );
    }
}
