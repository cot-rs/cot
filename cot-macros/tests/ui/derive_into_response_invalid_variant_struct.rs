use cot::response::IntoResponse;

#[derive(IntoResponse)]
enum MyResponse {
    A { field: u32 },
}

fn main() {}
