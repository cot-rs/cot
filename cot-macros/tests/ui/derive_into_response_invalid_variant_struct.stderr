error: only tuple variants with a single field are supported
 --> tests/ui/derive_into_response_invalid_variant_struct.rs:3:10
  |
3 | #[derive(IntoResponse)]
  |          ^^^^^^^^^^^^
  |
  = note: this error originates in the derive macro `IntoResponse` (in Nightly builds, run with -Z macro-backtrace for more info)
